rfd/0144-client-tools-updates.md

```diff
@@ -0,0 +1,320 @@
+---
+authors: Russell Jones (rjones@goteleport.com) and Bernard Kim (bernard@goteleport.com)
+state: draft
+---
+
+# RFD 0144 - Client Tools Updates
+
+## Required Approvers
+
+* Engineering: @sclevine && @bernardjkim && @r0mant
+* Product: @klizhentas || @xinding33
+* Security: @reedloden
+
+## What/Why
+
+This RFD describes how client tools (like `tsh` and `tctl`) can be kept up to
+date, either using automatic updates or self-managed updates.
+
+Keeping client tools updated helps with security (fixes for known security
+vulnerabilities are pushed to endpoints), bugs (fixes for resolved issues are
+pushed to endpoints), and compatibility (users no longer have to learn and
+understand [Teleport component
+compatibility](https://goteleport.com/docs/upgrading/overview/#component-compatibility)
+rules).
+
+## Details
+
+### Summary
+
+Client tools (like `tsh` and `tctl`) will automatically download and install
+the version of client tools recommended by the Teleport cluster.
+
+The `cluster_maintenance_config` resource will be updated to include a
+`tools_version` field to control the version of client tools advertised by the
+cluster. This resource will not be configurable on Teleport Cloud. It will be
+managed by the Teleport Cloud team.
+
+If cluster administrators want to opt-out of automatic updates and manage
+updates themselves, they will be able to watch the `cluster_maintenance_config`
+resource for changes and push out updates to endpoints manually.
+
+Inspiration drawn from https://go.dev/doc/toolchain.
+
+### Implementation
+
+#### Client tools
+
+##### Automatic updates
+
+The first time a user executes `tsh login` with a version of `tsh` that
```

Comment:
Should we plan to release client tools auto updates in a staged rollout?

Looking at how agent auto updates rollout has turned out, it might be worth considering a staged rollout where we release this feature to internal, trial, and team plans first. Once we've tested for a while, we can start rolling it out for enterprise.

If the feature is released with some bugs, I could see some users disabling the feature and never think about re-enabling it.

Reply:
That should be possible since `cluster_maintenance_config` is a per-cluster resource. I'll cover that part in a Cloud specific RFD.